### PR TITLE
[WIP] Fix --hostname-override being ignored when using cloud provider

### LINKED ISSUE #54482: --hostname-override ignored when --cloud-provider is specified
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig aws

**What happened**:
Trying to start kubelet with --hostname-override=ip-172-28-68-60 but still see in the logs:
```
Attempting to register node ip-172-28-68-60.eu-west-1.compute.internal
Unable to register node "ip-172-28-68-60.eu-west-1.compute.internal" with API server: nodes "ip-172-28-68-60.eu-west-1.compute.internal" is forbidden: node "ip-172-28-68-60" cannot modify node "ip-172-28-68-60.eu-west-1.compute.internal"
```
ps aux:
root      4610  3.3  7.7 404596 78320 ?        Ssl  12:58   0:00 /usr/bin/kubelet ...... --hostname-override=ip-172-28-68-60

**What you expected to happen**:
Hostname should be ip-172-28-68-60 instead of ip-172-28-68-60.eu-west-1.compute.internal

**How to reproduce it (as minimally and precisely as possible)**:
set
--cloud-provider=aws --hostname-override=ip-172-28-68-60
for kubelet

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.1", GitCommit:"f38e43b221d08850172a9a4ea785a86a3ffa3b3a", GitTreeState:"clean", BuildDate:"2017-10-11T23:27:35Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.1", GitCommit:"f38e43b221d08850172a9a4ea785a86a3ffa3b3a", GitTreeState:"clean", BuildDate:"2017-10-11T23:16:41Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration**: aws

- OS (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

- Kernel (e.g. `uname -a`): Linux ip-172-28-68-60 4.4.0-1038-aws #47-Ubuntu SMP Thu Sep 28 20:05:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by 2rs2ts**
This is an attempt to fix the behavior without changing the kubelet interface.
I believe that it would be better to allow overriding the hostname and the nodeName
separately via CLI args, and to use the hostname for looking up instances by their
DNS names, while using nodeName for its other normal purposes. However,
that sort of design discussion might take a while and we need this to be fixed now.

The solution we came up with is thus:

Keep track of what we want to set as the host/nodename and what the
cloud integration wants to set as separate things. The cloud integration
only cares about being able to identify the node as a cloud instance; it
doesn't actually care what you want to use as your node name.

Signed-off-by: Brett Kochendorfer <brett.kochendorfer@gmail.com>

Fixes #54482 #22984

**Special notes for your reviewer**: I believe that my implementation indicates
a code smell as I described above, so we didn't invest any time in updating the unit
tests to catch a regression, since I suspect the implementation may be subject to
change. After review we'll update the unit tests to catch a regression on this fix.
Hence the WIP.

```release-note
Fixed kubelet --hostname-override arg being ignored when using a cloud provider. See #22984, #54482
```

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -902,6 +902,7 @@ type Kubelet struct {
 	kubeletConfiguration kubeletconfiginternal.KubeletConfiguration
 
 	hostname        string
+	actualHostname  types.NodeName
```
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: If this is only used as the cloud provider id operations, maybe name it that instead? hostname and actualHostname raise eyebrows

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Sure, I'm amenable to any suggestions. I also would like to hear your thoughts on whether we should be overriding not the `hostname` but the `nodeName` through a CLI arg (or perhaps, whether we want to allow overriding either or both.) I thought it weird to have to add an extra piece of data at all, and the variable name we chose emerged as a symptom of that... it really is the _actual_ hostname, not whatever you have overridden it to

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Looks like there has been an expressed with for a nodename override CLI arg for a while now. See #22984

Comment (liggitt) [reply pkg/kubelet/kubelet.go]: There's a long and storied past to this.

In non-cloud environments, hostnames cannot always be auto-detected. --hostname-override was intended to let you override the hostname the node reports in those cases. 

Originally, the reported hostname and the nodeName were identical. That is no longer required to be the case (as of 1.5ish, I think). So we have two degrees of freedom currently:
* --hostname-override specified or unspecified
* cloud provider environment or non-cloud provider environment

As part of moving cloud providers out of process, we probably want to:
* allow both hostname and nodename to be specifiable via the command line
* insulate how the in-process cloud providers look up instance info (they likely shouldn't be affected by hostname override)
* possibly disallow setting hostname/nodeName in ways that disagree with the in-process cloud provider

Definitely need the table of current behavior when those options are set (and match, or mismatch), to know how we can get this straightened out in the least surprising and compatible way

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Ok, I think I understand. This is definitely made more complicated by the work of splitting out cloud stuff.

Given that we (meaning myself and my company) do have a need for the cloud-determined hostname to not override what we are using as nodeName, and given that we all would benefit from insulating the command line values from the cloud-determined value conceptually, perhaps my approach here is actually a good transition step, since it introduces the idea that these values should be tracked separately, and I just need to name the variable better. What do you think?

Comment (liggitt) [reply pkg/kubelet/kubelet.go]: > Given that we (meaning myself and my company) do have a need for the cloud-determined hostname to not override what we are using as nodeName

When using an in-process cloud provider, I'd expect the cloud provider to be the authority on both the hostname and node name.

> given that we all would benefit from insulating the command line values from the cloud-determined value conceptually

I actually think it is a mistake to allow setting hostname-override in a way that conflicts with the cloud-provider-determined values ("possibly disallow setting hostname/nodeName in ways that disagree with the in-process cloud provider")

By "insulate", I mostly meant that you shouldn't be able to set --hostname-override and have that influence how the in-process cloud provider code locates the node. If you're running an aws node foo, and you set --hostname-override=bar, I'd expect the in-process AWS cloud provider code to still be able to figure out you are on node "foo", and then prevent you from starting because you provided a conflicting overriding hostname.

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: So I suppose you are suggesting that we surface changing the node name first, then disallow changing the hostname in a way that conflicts with the hostname the cloud provider returns?

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: An additional thought: it is entirely possible to bind kubelet to a separate ENI in AWS, and that won't match the private DNS name. You can also have DNS records pointing at that address, and that _definitely_ won't match the private DNS name. 😄 We should consider that as we think of solutions.

Comment (liggitt) [reply pkg/kubelet/kubelet.go]: > So I suppose you are suggesting that we surface changing the node name first

Yes. This enables out of process cloud providers, and non-cloud-provider environments, and allows distinct control over both pieces of info.

> then disallow changing the hostname in a way that conflicts with the hostname the cloud provider returns?

Yes. Down that road lies pain.

Comment (liggitt) [reply pkg/kubelet/kubelet.go]: > it is entirely possible to bind kubelet to a separate ENI in AWS, and that won't match the private DNS name. You can also have DNS records pointing at that address, and that definitely won't match the private DNS name. 😄 We should consider that as we think of solutions.

that sounds like a prime use case for setting --hostname-override, --node-name, and not expecting the in-process AWS cloud provider to work.

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Why would it be reasonable not to expect the cloud provider to work? Unless I'm ignorant of some feature set that cloud providers are supposed to have, there's no reason why you shouldn't be able to identify the instance kubelet is running on (if not automatically, then manually.)

My colleague @bkochendorfer and I were talking through this: it should be entirely possible, if you're on AWS, to get your own instance ID and your own private DNS name from the metadata endpoint. You can totally ignore whatever you passed in for `--hostname-override`. But if for some reason you can't get that information, perhaps it would be good to provide it at the CLI. And that concept should be referred to as a cloud provider name, or ID, or something of the sort.

It may be wise to stop using `types.NodeName` for the arguments to provider methods like `instances.CurrentNodeName` and `instances.ExternalID` and instead have a type, something like `CloudProviderName` or whatever, so that the concepts aren't getting mixed up.

Now you can override your hostname/nodename to be the IP address of that ENI, or a DNS record pointing at that address, or whatever you like. And you can also pass in whatever cloud identifier your provider needs to work, if it somehow can't get it itself.

```
--hostname-override 10.1.2.3 \
--nodename-override node01 \
--cloud-provider-name-override ip-10-4-5-6.us-west-2.compute.internal
```

So, in other words, I believe that you should _not_ disallow changing the hostname to conflict with whatever your cloud provider wants to use as the identifier for finding the instance in the cloud.

What do you think about this? I may be missing a lot of details, but I think this forms a complete solution:

1. `hostname` can be used for routing requests to the kubelet/kube-proxy
2. `NodeName` can be used for things like `kubectl get node node01`
3. `CloudProviderName` can be used for things like EC2 API calls

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Thinking through it more we don't believe that being able to provide the cloud provider name at startup is useful. That sort of thing should be discoverable at startup using instance metadata. I am only familiar with AWS but I'd be surprised if other clouds wouldn't have a good way to figure this out.

We do still think that there should be a separate type and all for this. How the cloud provider identifies the instance is different from how kubernetes wants to identify the Node and route requests to it.

Comment (liggitt) [reply pkg/kubelet/kubelet.go]: > Why would it be reasonable not to expect the cloud provider to work?

In cloud environments, you need an authority on the following things (and the mapping between them):
* Node machine IDs (for things like ec2 API calls)
* Node hostnames (for determining whether a node should be given a serving cert for a particular hostname, etc)
* Node API object names (for node controller, attach/detach controller, persistent volume provisioners, etc to reconcile cloud provider state to nodes and their associated resources)

The kubelets cannot be the authority on those things:
* before a kubelet has reported state, no other components can know anything about what nodes exist
* it lets a kubelet request or interfere with resources (hostnames, volumes, etc) it shouldn't have access to

> I believe that you should not disallow changing the hostname to conflict with whatever your cloud provider wants to use as the identifier for finding the instance in the cloud.

That was only a statement for the compiled-in cloud provider components.

> I think this forms a complete solution:
> * hostname can be used for routing requests to the kubelet/kube-proxy
> * NodeName can be used for things like kubectl get node node01
> * CloudProviderName can be used for things like EC2 API calls

I agree those are the three pieces of info needed, but if you want non-standard mappings between those, I'd rather see that satisfied by using out of process cloud provider components driven by an authoritative mapping source of your choice, and just make the kubelet be told its hostname and nodename, instead of trying to make the compiled-in cloud provider components deal with a variety of mappings.

Comment (2rs2ts) [reply pkg/kubelet/kubelet.go]: Hmm, ok, I see what you mean. Do you think that the cloud-controller-manager would be the place to implement this sort of mapping?

**Comment by 2rs2ts** [comment]
Closing because this was contentious
